entity-service/src/main/resources/configs/common/application.conf

```diff
@@ -21,6 +21,12 @@ attribute.service.config = {
   port = ${?ATTRIBUTE_SERVICE_PORT_CONFIG}
 }
 
+entity.service.change.skip.attributes = [
+  {
+    scope = API
+    attributes = ["last_called_millis"]
```

Comment:
I wanted to do that initially, but if we map it to attribute ID then we have to add the dependency to the CachingAttributeClient to EntityChangeEventGenerator. Because then we need to resolve these attributes and their paths. Is there any other way to achieve this without CachingAttributeClient? Adding the dependency of AttributeClient seems overkill.

Reply:
Not necessarily needed. The attribute client helps go from an attribute ID to its metadata - here, you need to go from an entity subpath to an attribute ID. EQS already has a map of data to do this in the other direction, and that's what you'd need access to.

Reply:
I dont think i understand this correctly. Lets say we define the skip map something like this
```
entity.service.change.skip.attributes = [
        {
               scope = API,
               attributeId = API.lastCalledTimestamp
               // or
              subDocPath = attributes.last_called_millis (though it's same as above just because of subdoc path we are adding attributes.)
        
        }
]
```

Now in case one, where we are defining attributeId, we have to integrate with attributesClient because we can't be sure that the given attribute will be part of the event service config. It can come from attribute service also. And EventGenerator needs the exact doc store path.
Now the second case with subDocPath is similar to what I have already defined. It's better to add the attribute name directly to skip some translation logic (to remove attributes from subdoc path).

Is my understanding correct? or am I going in a completely wrong direction here?

Reply:
I'm suggesting here that we define it like:
```
// Just attribute ids
entity.service.change.skip.attributes = ["API.lastCalledTimestamp", "API.xyz"]
```

Now, when the change event generator is handed an entity, it can basically take this https://github.com/hypertrace/entity-service/blob/6f000f36d732bcb49ce60b00adc3afdbe71ee8a3/entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityAttributeMapping.java#L59-L63

And just added the inverse function - `getAttributeIdFromDocStorePath`. This should allow each value in the entity to be mapped to an attribute ID, which can then be checked against the list. No attribute lookup needed. There may be a few fields without attributes, but that's purely tech debt (because we stopped defining separate entity attributes years ago), so we should just default to not support skipping them and if we run into the scenario where we need to, it's forcing function to clean up tech debt.

Does that make sense? That's basically what I was thinking at a high level.

Reply:
`calculateDocStorePathFromAttributeId` actually makes a call to the attribute service Client. So if I understand this correctly, what you are suggesting is for each attribute in the entity, we will call `getAttributeIdFromDocStorePath` which will return list of AttributeIds and then filter out based on configuration present. 

But the problem still remains the same, even though there will be a few attributes that are not defined, it still needs to make the call to the attribute client to verify all the possible places where we can define the attributes.

If we really want to go with AttributeId, i think we should go with something like this:
```
// Just attribute ids
entity.service.change.skip.attributes = ["API.lastCalledTimestamp", "API.xyz"]

private EntityAttributeMapping attributeMapping
public EntityChangeEventGenerator(EntityAttributeMapping attributeMapping) {
attributeMapping = attributeMapping;
}

public void generateEvent(Entity entity) {
 attributePaths = skipAttributes.forEach(attribute -> attributeMapping.getDocStorePathByAttributeId(attribute)).collect(Collectors.toList());
// . . . 
Filter out the attributes from entity and then check the difference.
// . . . .
}
```

Reply:
Discussed a bit offline, just 2 bits to add here:
- Making a call to attribute client shouldn't really be an issue if perf is your concern here. It's a caching client and already called anyway (assuming we have called EQS), so there's no remote calls being made here.
- the above is fine - whether we start with the subdoc path and translate it to an id, or give all skipped ids and translate to subdoc paths isn't a huge deal given the relevant quantities. Only thing missing there is that you've lost your concept of scope when you got those paths back (e.g. I mapped API.xyz to the subdoc path xyz, but for all I know the entity is of type service which _also_ has a subdoc path xyz that hasn't been mapped to skipping). So there's going to be a bit more to do, but not much.

Reply:
I am going ahead with the above approach. In this way, I don't have to translate all the available attributes in an entity.